docs/developing.rst

```diff
@@ -742,3 +742,246 @@ Note the call to :func:`toil.job.Job.encapsulate` creates the
+invoke like any Python script::
+
+   $ cd my_project
+   $ ls
+   userScript.py …
+   $ ./userScript.py …
+
+This assumes that your script is executable and contains a *shebang*, i.e. a
+line of the form
+
+::
+
+   #!/usr/bin/python
+
+or equivalent. The script can have dependencies, as long as those are installed
+on the machine, either globally, in a user-specific location or in a
+virtualenv. In the latter case, the virtualenv must of course be active when
+you run the user script.
+
+If, however, you want to run your workflow in a distributed environment, on
+multiple worker machines, either in the cloud or on a bare-metal cluster, your
+script needs to be made available to those other machines. If your script
+imports other modules, those modules also need to be made available on the
+workers. Toil can automatically do that for you, with a little help on your
+part. We call this feature ``hot-deployment`` of a workflow.
+
+
+Hot-deployment without dependencies
+-----------------------------------
+
+Your script has no additional dependencies, it imports only modules that are
+shipped with Python. Toil will take your script, store it in the job store and
+just before the jobs in your script are about to be run on a worker machine,
+your script will be saved to a temporary directory on the worker and loaded
+into the Python interpreter from there::
+
+   $ cd my_project
+   $ ls
+   userScript.py
+   $ ./userScript.py --batchSystem=mesos …
+   
+
+
+Hot-deployment with sibling modules
+-----------------------------------
+
+Your script only imports modules that are siblings of the user script::
+
+   $ cd my_project
+   $ ls
+   userScript.py utilities.py
+   $ ./userScript.py …
+
+where ``userScript.py`` imports additional functionality from ``utilities.py``.
+Toil detects that userScript.py has sibling modules—independent of whether they
+are actually imported by the user script—and copies those to the workers as
+well before invoking jobs. This is a decent way of organizing the source code
+of reasonably complicated workflows.
+
+
+Hot-deploying a package hierarchy
+---------------------------------
+
+Remember that in Python, a package is a directory containing one or more
+``.py`` files, one of which must be called ``__init__.py``, and optionally
+other packages. For more involved workflows that contain a significant amount
+of code, this is the recommended way of organizing the source. Because we use a
+package hierarchy, we can't really refer to the user script as such, we now
+call it the *user module*. It's simply one of the modules in the package
+hierarchy and we need to inform Toil about that by invoking it with Python `-m`
+option. That way Toil will be able to identify the entire set of modules
+belonging to the workflow and copy the right ones to each worker. Let's say we
+organized our sample project as follows::
+
+   $ cd my_project
+   $ tree
+   .
+   ├── bar
```

Comment:
I agree that not being genomics-specific is probably the right call for this doc example.